analytic_engine/src/sst/parquet/meta_data.rs

```diff
@@ -45,89 +58,201 @@ pub enum Error {
 
 define_result!(Error);
 
-const DEFAULT_BLOOM_FILTER_VERSION: u32 = 0;
+const DEFAULT_FILTER_VERSION: u32 = 0;
 
-#[derive(Debug, Clone, PartialEq, Eq, Default)]
-pub struct RowGroupBloomFilter {
-    // The column filter can be None if the column is not indexed.
-    column_filters: Vec<Option<Bloom>>,
+/// Filter can be used to test whether an element is a member of a set.
+/// False positive matches are possible if space-efficient probabilistic data
+/// structure are used.
+trait Filter: fmt::Debug {
+    /// Check the key is in the bitmap index.
+    fn contains(&self, key: &[u8]) -> bool;
+
+    /// Serialize the bitmap index to binary array.
+    fn to_bytes(&self) -> Vec<u8>;
+
+    /// Deserialize the binary array to bitmap index.
+    fn from_bytes(buf: Vec<u8>) -> Result<Self>
+    where
+        Self: Sized;
+}
+
+/// Filter based on https://docs.rs/xorfilter-rs/latest/xorfilter/struct.Xor8.html
+#[derive(Default)]
+struct Xor8Filter {
+    xor8: Xor8,
+}
+
+impl fmt::Debug for Xor8Filter {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.write_str("XorFilter")
+    }
+}
+
+impl Filter for Xor8Filter {
+    fn contains(&self, key: &[u8]) -> bool {
+        self.xor8.contains(key)
+    }
+
+    fn to_bytes(&self) -> Vec<u8> {
+        self.xor8.to_bytes()
+    }
+
+    fn from_bytes(buf: Vec<u8>) -> Result<Self>
+    where
+        Self: Sized,
+    {
+        Xor8::from_bytes(buf)
+            .context(ParseXor8Filter)
+            .map(|xor8| Self { xor8 })
+    }
 }
 
-impl RowGroupBloomFilter {
-    pub fn with_num_columns(num_columns: usize) -> Self {
+pub struct RowGroupFilterBuilder {
+    builders: Vec<Option<Xor8Builder>>,
+}
+
+impl RowGroupFilterBuilder {
+    pub(crate) fn with_num_columns(num_col: usize) -> Self {
         Self {
-            column_filters: vec![None; num_columns],
+            builders: vec![None; num_col],
         }
     }
 
-    /// Accrue the data belonging to one column.
-    ///
-    /// Caller should ensure the `column_idx` is in the range.
-    pub fn accrue_column_data(&mut self, column_idx: usize, data: &[u8]) {
-        if self.column_filters[column_idx].is_none() {
-            self.column_filters[column_idx] = Some(Bloom::default());
+    pub(crate) fn add_key(&mut self, col_idx: usize, key: &[u8]) {
```

Comment:
How about defining a trait `FilterBuilder`?

Reply:
Good idea, I plan to move Filter related code to `src/meta_data.rs`, currently it's in wrong place.

I left a comment to add FilterBuilder trait when refactor this.